File: python/base_plugin.py
import threading
import time
from .plugin_sdk import PluginSDK
from .plugin_info import PluginInfo
from .command_result import CommandResult
from .grpc_client import GrpcClient

class BasePlugin(PluginSDK):
    """基础插件实现类"""
    
    def __init__(self):
        self.config = None
        self.info = PluginInfo()
        self.client = None
        self.running = False
        self.heartbeat_thread = None
        
    def initialize(self, config):
        """
        初始化插件
        :param config: 插件配置
        :return: 是否初始化成功
        """
        self.config = config
        
        # 设置插件基本信息
        self.info.set_id(config.get_plugin_id())
        self.info.set_name(config.get_plugin_name())
        self.info.set_version(config.get_plugin_version())
        self.info.set_type(config.get_plugin_type())
        
        # 创建 gRPC 客户端
        self.client = GrpcClient(config.get_server_host(), config.get_server_port())
        
        return True
    
    def start(self):
        """
        启动插件
        :return: 是否启动成功
        """
        if self.running:
            return True
            
        # 注册插件
        response = self.client.register_plugin(
            self.info.get_name(),
            self.info.get_version(),
            self.info.get_type(),
            self.info.get_description(),
            self.config.get_config("plugin.host"),
            int(self.config.get_config("plugin.port"))
        )
        
        if not response.success:
            print(f"插件注册失败: {response.message}")
            return False
            
        # 设置插件ID
        self.info.set_id(response.plugin_id)
        self.config.set_plugin_id(response.plugin_id)
        
        # 启动心跳线程
        self.running = True
        self.heartbeat_thread = threading.Thread(target=self._heartbeat_loop)
        self.heartbeat_thread.daemon = True
        self.heartbeat_thread.start()
        
        return True
    
    def stop(self):
        """
        停止插件
        :return: 是否停止成功
        """
        if not self.running:
            return True
            
        self.running = False
        
        if self.heartbeat_thread:
            self.heartbeat_thread.join(timeout=2.0)
            
        if self.client:
            self.client.stop_plugin(self.info.get_id())
            self.client.close()
            
        return True
    
    def get_info(self):
        """
        获取插件信息
        :return: 插件信息
        """
        return self.info
    
    def execute_command(self, command, params):
        """
        执行插件命令
        :param command: 命令
        :param params: 参数
        :return: 命令执行结果
        """
        # 子类应该重写此方法
        return CommandResult(False, None, "未实现的命令")
    
    def handle_message(self, message):
        """
        处理来自主应用的消息
        :param message: 消息内容
        :return: 处理结果
        """
        # 子类应该重写此方法
        return "未处理的消息"
    
    def _heartbeat_loop(self):
        """心跳循环"""
        while self.running:
            try:
                response = self.client.send_heartbeat(
                    self.info.get_id(),
                    self.info.get_status() or "RUNNING"
                )
                # 可以处理心跳响应
            except Exception as e:
                print(f"心跳发送失败: {str(e)}")
            
            # 每5秒发送一次心跳
            time.sleep(5)

File: cpp/base_plugin.cpp
#include "base_plugin.h"
#include <chrono>
#include <iostream>
#include <grpcpp/grpcpp.h>
#include "plugin_service.grpc.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;
using plugin::PluginService;
using plugin::PluginRegistration;
using plugin::RegistrationResponse;
using plugin::HeartbeatRequest;
using plugin::HeartbeatResponse;
using plugin::StatusRequest;
using plugin::StatusResponse;
using plugin::CommandRequest;
using plugin::CommandResponse;
using plugin::StopRequest;
using plugin::StopResponse;

BasePlugin::BasePlugin() : running(false) {
}

BasePlugin::~BasePlugin() {
    stop();
}

bool BasePlugin::initialize(const PluginConfig& config) {
    this->config = config;
    
    // 设置插件基本信息
    info.setId(config.getPluginId());
    info.setName(config.getPluginName());
    info.setVersion(config.getPluginVersion());
    info.setType(config.getPluginType());
    
    return true;
}

bool BasePlugin::start() {
    if (running) {
        return true;
    }
    
    // 创建 gRPC 通道
    std::string serverAddress = config.getServerHost() + ":" + std::to_string(config.getServerPort());
    channel = grpc::CreateChannel(serverAddress, grpc::InsecureChannelCredentials());
    stub = PluginService::NewStub(channel);
    
    // 注册插件
    PluginRegistration request;
    request.set_name(info.getName());
    request.set_version(info.getVersion());
    request.set_type(info.getType());
    request.set_description(info.getDescription());
    request.set_host(config.getConfig("plugin.host"));
    request.set_port(std::stoi(config.getConfig("plugin.port")));
    
    RegistrationResponse response;
    ClientContext context;
    Status status = stub->RegisterPlugin(&context, request, &response);
    
    if (!status.ok() || !response.success()) {
        std::cerr << "插件注册失败: " << response.message() << std::endl;
        return false;
    }
    
    // 设置插件ID
    info.setId(response.plugin_id());
    
    // 启动心跳线程
    running = true;
    heartbeatThread = std::thread(&BasePlugin::heartbeatLoop, this);
    
    return true;
}

bool BasePlugin::stop() {
    if (!running) {
        return true;
    }
    
    running = false;
    
    if (heartbeatThread.joinable()) {
        heartbeatThread.join();
    }
    
    if (stub) {
        StopRequest request;
        request.set_plugin_id(info.getId());
        
        StopResponse response;
        ClientContext context;
        Status status = stub->StopPlugin(&context, request, &response);
    }
    
    return true;
}

PluginInfo BasePlugin::getInfo() const {
    return info;
}

CommandResult BasePlugin::executeCommand(const std::string& command, const std::map<std::string, std::string>& params) {
    // 子类应该重写此方法
    return CommandResult(false, "", "未实现的命令");
}

std::string BasePlugin::handleMessage(const std::string& message) {
    // 子类应该重写此方法
    return "未处理的消息";
}

void BasePlugin::heartbeatLoop() {
    while (running) {
        try {
            HeartbeatRequest request;
            request.set_plugin_id(info.getId());
            request.set_status_info(info.getStatus().empty() ? "RUNNING" : info.getStatus());
            
            HeartbeatResponse response;
            ClientContext context;
            Status status = stub->Heartbeat(&context, request, &response);
            
            if (!status.ok()) {
                std::cerr << "心跳发送失败: " << status.error_message() << std::endl;
            }
        } catch (const std::exception& e) {
            std::cerr << "心跳发送异常: " << e.what() << std::endl;
        }
        
        // 每5秒发送一次心跳
        std::this_thread::sleep_for(std::chrono::seconds(5));
    }
}